WordCamp Payments: Send organizers reminder about pending invoices.

vedanshujain opened the pull request:
Sometimes sponsor invoices do not get paid on time. This patch adds a reminder, currently configured on 30th, 45th and 60th day after invoice is approved.

vedanshujain committed:
WordCamp Payments: Send organizers reminder about pending invoices.

Sometimes sponsor invoices do not get paid on time. This patch adds a reminder, currently configured on 30th, 45th and 60th day after invoice is approved.

coreymckrill commented:
I'm curious to hear what @iandunn thinks as well, but my first thought is maybe we should schedule one cron job once a day that grabs all the unpaid invoices from the invoice index table and sends reminder emails for each one that meets the criteria. Instead of automatically scheduling separate cron jobs for every invoice...

iandunn commented on public_html/wp-content/plugins/wordcamp-payments/includes/sponsor-invoice.php:
```diff
@@ -419,8 +419,89 @@ function save_invoice( $post_id, $post ) {
+
+	$edit_link = get_site_url() . "/wp-admin/post.php?post=$invoice_post_id&action=edit";
+
+	$mail = <<<HTML
```
What do you think about using something like this instead of heredoc?

```php
$message = str_replace( "\t", '', "
	The invoice for `{$invoice->post_title}` $status_message

	You can view the invoice and its status any time at:

	$invoice_url
	$attachment_message

	If you have any questions, please reply to let us know."
);
```

That's from `WordCamp\Budgets_Dashboard\Sponsor_Invoices\notify_organizer_status_changed()`. I like it because it doesn't break the indentation pattern. Not a big deal, though.

vedanshujain replied:
:+1:

iandunn commented on public_html/wp-content/plugins/wordcamp-payments/includes/sponsor-invoice.php:
```diff
@@ -419,8 +419,89 @@ function save_invoice( $post_id, $post ) {
+
 /**
  * Define columns for the Vendor Payments screen.
  *
```
I noticed a few things that I think PHPCS will warn about.

vedanshujain commented on public_html/wp-content/plugins/wordcamp-payments/includes/sponsor-invoice.php:
```diff
@@ -419,8 +419,85 @@ function save_invoice( $post_id, $post ) {
+			wp_schedule_single_event( time() + DAY_IN_SECONDS * 30, 'send_invoice_pending_reminder', array( $post_id ) );
+			update_post_meta( $post_id, 'reminder_schedule_start_time', time() );
+		}
+	}
```
I have switched to a single cron job that you and @coreymckrill suggested. That does makes more sense, and since we have a index table its also good in terms of performance. Resolving this discussion now.

vedanshujain replied:
Also for older post, I am assuming the current time to be the time reminder sent at, because I do not think we do have any better information that we can use to assume.

vedanshujain commented on public_html/wp-content/plugins/wordcamp-payments/includes/sponsor-invoice.php:
```diff
@@ -419,8 +419,85 @@ function save_invoice( $post_id, $post ) {
 		}
 	}
+
+	$should_schedule_invoice_reminder = apply_filters( 'should_schedule_invoice_reminder', true );
```
It was a just in case. I have removed it in the re-implementation though

iandunn replied:
👍

vedanshujain commented on public_html/wp-content/plugins/wordcamp-payments/includes/sponsor-invoice.php:
```diff
@@ -419,8 +419,89 @@ function save_invoice( $post_id, $post ) {
+	$mail = <<<HTML
+Howdy organizers,
+<br>
+Looks like the invoice <a href='$edit_link' target='_blank'>{$invoice->post_title}</a> is still unpaid. If you still expect the sponsor to pay this invoice, please contact them to find out when we should expect payment. If this invoice needs to be cancelled, please email support@wordcamp.org.
```
fixed

iandunn replied:
👍

vedanshujain commented on public_html/wp-content/plugins/wordcamp-payments/includes/sponsor-invoice.php:
```diff
@@ -419,8 +419,89 @@ function save_invoice( $post_id, $post ) {
+	$mail = <<<HTML
+Howdy organizers,
+<br>
+Looks like the invoice <a href='$edit_link' target='_blank'>{$invoice->post_title}</a> is still unpaid. If you still expect the sponsor to pay this invoice, please contact them to find out when we should expect payment. If this invoice needs to be cancelled, please email support@wordcamp.org.
```
I am inclined to keep the link, because that way organizer will not have to search for it. Although I can remove it as well, if you think.

iandunn replied:
Yeah, keeping it is probably the more practical option.

vedanshujain commented on public_html/wp-content/plugins/wordcamp-payments/includes/sponsor-invoice.php:
```diff
@@ -419,8 +419,85 @@ function save_invoice( $post_id, $post ) {
+	if ( 'wcbsi_approved' !== $status ) {
+		// Looks like Invoice is not pending anymore. Nothing to do.
+		return;
+	}
```
Oh did you maybe accidentally misread the equality sign. You are correct that `wcbsi_approved` means that it is sent to the sponsor and not paid. If the status is not this then its either paid, or changed into draft again, and in both of these cases we would want to return. This is what you mean right?

iandunn replied:
Ah, that makes sense, thanks!

vedanshujain commented:
@iandunn @coreymckrill  can you look at this again. Made some changes in approach to have single cron process reminders

iandunn commented on public_html/wp-content/plugins/wordcamp-payments-network/includes/sponsor-invoices-dashboard.php:
```diff
@@ -461,3 +464,107 @@ function delete_index_row( $invoice_id ) {
+		switch_to_blog( $blog_id );
+
+		$invoice         = get_post( $invoice_id );
+		$invoice_sent_at = get_post_meta( $invoice_id, 'Sent at', true );
```
This can also be retrieved via `$invoice->{'Sent at'}`, which is a bit more readable. Same for `$last_reminder` and `$invoice_sent_at` below.

iandunn commented on public_html/wp-content/plugins/wordcamp-payments-network/includes/sponsor-invoices-dashboard.php:
```diff
@@ -461,3 +464,107 @@ function delete_index_row( $invoice_id ) {
+		);
+
+		if ( $reminder_step > count( $reminder_schedule ) ) {
+			// At this point 2 months have been passed. Maybe send a mail to Community support at this point.
```
It sounds like Andrea wants to do this. This function is already pretty long, so maybe at this point would should modularize it a bit. Maybe in this condition you could call a new function that handles everything related to sending the email to central. Then you could create another function that contains all the code below that's related to sending the email to the organizer?

vedanshujain replied:
thanks for the suggestion. done.

coreymckrill commented on public_html/wp-content/plugins/wordcamp-payments-network/includes/sponsor-invoices-dashboard.php:
```diff
@@ -461,3 +464,107 @@ function delete_index_row( $invoice_id ) {
+			update_post_meta( $invoice_id, 'Backfilled Sent at', true );
+			restore_current_blog();
+			continue;
+		}
```
Just a thought: you could get more accurate sent dates for the backfill using our QBO API client, like is used in the Sponsor Invoices report:

https://meta.trac.wordpress.org/browser/sites/trunk/wordcamp.org/public_html/wp-content/plugins/wordcamp-reports/classes/report/class-sponsor-invoices.php#L274

vedanshujain replied:
Since its not a very strict requirement, and we are anyways defaulting the invoice after 60 days since WordCamp, I'd rather not do this extra work which we are not gonna use in future. Do you think there perhaps could be a good reason to prefill accurately that I am not thinking of?

coreymckrill replied:
> Do you think there perhaps could be a good reason to prefill accurately that I am not thinking of?

Nope. If you don't think the accuracy is necessary here, then I'm not concerned about it either :)

coreymckrill commented on public_html/wp-content/plugins/wordcamp-payments/includes/sponsor-invoice.php:
```diff
@@ -419,6 +419,13 @@ function save_invoice( $post_id, $post ) {
+	if ( 'wcbsi_approved' === $post->post_status ) {
+		$invoice_sent_at = get_post_meta( $post_id, 'Sent at', true );
+		if ( empty( $invoice_sent_at ) ) {
+			update_post_meta( $post_id, 'Sent at', true );
```
I think the value here should be `time()` instead of `true`?

vedanshujain replied:
fixed

coreymckrill commented on public_html/wp-content/plugins/wordcamp-payments-network/includes/sponsor-invoices-dashboard.php:
```diff
@@ -461,3 +464,107 @@ function delete_index_row( $invoice_id ) {
+
+		$author = get_user_by( 'ID', $invoice->post_author );
+		wp_mail(
+			array( $author->user_email ),
```
Should the recipient be the invoice creator, or maybe the main email address for the WordCamp? I'm thinking the latter might be better because often the main email address forwards to the whole organizing team.

iandunn replied:
Maybe send it to the invoice creator (since they're primarily responsible), but also CC the lead organizer or main address (in case the invoice creator has flaked out) ?

vedanshujain replied:
added a cc to lead organizer.

Commit:
WordCamp Misc: Disable Slack error notifications in local environments.

Local environments don't the credential to send to Slack, and have access to the local error log, so the Slack notifications aren't necessary and don't work.

git-svn-id: https://meta.svn.wordpress.org/sites/trunk/wordcamp.org@8137 74240141-8908-4e6f-9713-ba540dce6ec7

Commit:
WordCamp Helpers: Alias `log` function to improve readability.

git-svn-id: https://meta.svn.wordpress.org/sites/trunk/wordcamp.org@8138 74240141-8908-4e6f-9713-ba540dce6ec7

Commit:
WordCamp Post Type: Move `Event_Loader` require to plugin loader.

The class is needed for both `WordCamp_Loader` and `Meetup_Loader`, so it should be included higher up the chain, rather than relying on the assumption that `WordCamp_Loader` will load it before `Meetup_Loader` needs it.

This also makes it easier for other plugins to include `WordCamp_Loader`, in order to access it's public methods.

git-svn-id: https://meta.svn.wordpress.org/sites/trunk/wordcamp.org@8139 74240141-8908-4e6f-9713-ba540dce6ec7

Commit:
WordCamp Post Type: Disable Slack notifications in local environments.

git-svn-id: https://meta.svn.wordpress.org/sites/trunk/wordcamp.org@8140 74240141-8908-4e6f-9713-ba540dce6ec7

Commit:
WordCamp Payments: Prevent sending invoices before contract is signed.

git-svn-id: https://meta.svn.wordpress.org/sites/trunk/wordcamp.org@8141 74240141-8908-4e6f-9713-ba540dce6ec7

Commit:
Plugins: Remove unused Crowdsignal plugin.

The `[polldaddy]` shortcode has existed in Jetpack for a long time now, and there aren't any sites using the dedicated plugin anymore. 


git-svn-id: https://meta.svn.wordpress.org/sites/trunk/wordcamp.org@8142 74240141-8908-4e6f-9713-ba540dce6ec7

Commit:
WordCamp Misc: Ignore known Super Cache `no such file` errors.

git-svn-id: https://meta.svn.wordpress.org/sites/trunk/wordcamp.org@8143 74240141-8908-4e6f-9713-ba540dce6ec7

Commit:
WordCamp Network Plugins: Remove uninstalled Crowdsignal entry.

The plugin was removed in r8142, so it is no longer needed here.

git-svn-id: https://meta.svn.wordpress.org/sites/trunk/wordcamp.org@8144 74240141-8908-4e6f-9713-ba540dce6ec7

vedanshujain commented on public_html/wp-content/plugins/wordcamp-payments-network/includes/sponsor-invoices-dashboard.php:
```diff
@@ -461,3 +464,107 @@ function delete_index_row( $invoice_id ) {
+		$author = get_user_by( 'ID', $invoice->post_author );
+		wp_mail(
+			array( $author->user_email ),
+			"Pending invoice: $invoice->post_title",
```
done

Commit:
WordCamp Payments: Ensure super admins can update payment requests

In r8136 the button to save changes to vendor payment and reimbursement
requests was hidden from all users when no documentation files were attached,
but it should have still been available to super admins.



git-svn-id: https://meta.svn.wordpress.org/sites/trunk/wordcamp.org@8164 74240141-8908-4e6f-9713-ba540dce6ec7

vedanshujain committed:
WCPT: Disable new meetup organizer notification.

Disabling the new organizer notification because its a lot of work, and not always possible to find someone's wporg username from their meetup profile.



git-svn-id: https://meta.svn.wordpress.org/sites/trunk/wordcamp.org@8193 74240141-8908-4e6f-9713-ba540dce6ec7

vedanshujain commented:
this was merged.
